File: ios/LivenessDetection/LivenessView.swift
import UIKit

/// Embeddable liveness view. Add to any view hierarchy, assign `delegate`,
/// optionally tune `config` / `sounds` / `modelUrl`, then call `start()`.
///
/// - Full-bleed: add with autoresizing or constraints filling the container.
/// - Embed in a slot (e.g. a circular UI region): set `config.showInstructions = false`
///   so the host owns text and buttons. Sounds keep playing.
///
/// Camera permission must be granted by the host before calling `start()`.
@objc public final class LivenessView: UIView, LivenessDetectorDelegate {

  private let previewView = UIView()
  private let overlayView = LivenessOvalOverlayView()
  private let instructionLabel = UILabel()
  private let posHintLabel = UILabel()
  private let hintView = LivenessHintView()

  private var detector: LivenessDetector?

  /// Tuning + UI options. Reassign at any time to update theme/shape/visibility.
  @objc public var config: LivenessConfig = LivenessConfig() {
    didSet { applyConfig() }
  }
  /// Per-step sounds (left/blink/right/nod/mouth/good/capture).
  @objc public var sounds: LivenessSoundOptions?
  /// Override the MediaPipe Face Landmarker model URL. Defaults to Google's hosted float16 model.
  @objc public var modelUrl: String?
  /// Callbacks for the host. Invoked on the main thread.
  @objc public weak var delegate: LivenessDetectorDelegate?

  override public init(frame: CGRect) {
    super.init(frame: frame)
    setup()
  }

  public required init?(coder: NSCoder) {
    super.init(coder: coder)
    setup()
  }

  private func setup() {
    backgroundColor = .black

    previewView.translatesAutoresizingMaskIntoConstraints = false
    overlayView.translatesAutoresizingMaskIntoConstraints = false
    instructionLabel.translatesAutoresizingMaskIntoConstraints = false
    posHintLabel.translatesAutoresizingMaskIntoConstraints = false
    hintView.translatesAutoresizingMaskIntoConstraints = false

    instructionLabel.textColor = .white
    instructionLabel.textAlignment = .center
    instructionLabel.font = UIFont.systemFont(ofSize: 17, weight: .semibold)
    instructionLabel.text = "Position your face in the oval"

    posHintLabel.textColor = UIColor(red: 1, green: 59/255, blue: 59/255, alpha: 1)
    posHintLabel.textAlignment = .center
    posHintLabel.font = UIFont.systemFont(ofSize: 13, weight: .medium)
    posHintLabel.isHidden = true

    addSubview(previewView)
    addSubview(overlayView)
    addSubview(instructionLabel)
    addSubview(posHintLabel)
    addSubview(hintView)

    NSLayoutConstraint.activate([
      previewView.topAnchor.constraint(equalTo: topAnchor),
      previewView.bottomAnchor.constraint(equalTo: bottomAnchor),
      previewView.leadingAnchor.constraint(equalTo: leadingAnchor),
      previewView.trailingAnchor.constraint(equalTo: trailingAnchor),

      overlayView.topAnchor.constraint(equalTo: topAnchor),
      overlayView.bottomAnchor.constraint(equalTo: bottomAnchor),
      overlayView.leadingAnchor.constraint(equalTo: leadingAnchor),
      overlayView.trailingAnchor.constraint(equalTo: trailingAnchor),

      instructionLabel.topAnchor.constraint(equalTo: topAnchor, constant: 18),
      instructionLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
      instructionLabel.trailingAnchor.constraint(equalTo: trailingAnchor),

      posHintLabel.centerXAnchor.constraint(equalTo: centerXAnchor),
      posHintLabel.centerYAnchor.constraint(equalTo: centerYAnchor, constant: 40),

      hintView.centerXAnchor.constraint(equalTo: centerXAnchor),
      hintView.centerYAnchor.constraint(equalTo: centerYAnchor, constant: -80),
      hintView.widthAnchor.constraint(equalToConstant: 52),
      hintView.heightAnchor.constraint(equalToConstant: 52),
    ])

    applyConfig()
  }

  private func applyConfig() {
    overlayView.applyStyle(config)
    let show = config.showInstructions
    instructionLabel.isHidden = !show
    hintView.isHidden = !show
    if !show { posHintLabel.isHidden = true }
    posHintLabel.textColor = config.progressErrorColor
  }

  /// Begin liveness. Camera permission must be granted by the host.
  @objc public func start() {
    stop()
    if config.showInstructions {
      instructionLabel.text = "Position your face in the oval"
    }
    let det = LivenessDetector(config: config, delegate: self, sounds: sounds)
    detector = det
    det.startLiveness(previewView: previewView, useFrontCamera: true, modelUrl: modelUrl ?? LivenessDetector.defaultModelURL)
  }

  /// Stop detection and release the camera. Safe to call repeatedly.
  @objc public func stop() {
    detector?.stop()
    detector = nil
  }

  // ── LivenessDetectorDelegate (internal — updates UI then forwards) ────────

  public func onChallengeChanged(stepIndex: Int, stepLabel: String) {
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      if self.config.showInstructions {
        self.instructionLabel.text = stepLabel
      }
      if stepIndex >= 0 {
        self.overlayView.setProgress(stepIndex)
        self.overlayView.setStepDots(activeIndex: stepIndex)
        self.hintView.setHint(stepLabel: stepLabel)
      } else {
        self.overlayView.setProgress(5)
        self.overlayView.setStepDots(activeIndex: 5)
        self.hintView.setHint(stepLabel: nil)
      }
      self.delegate?.onChallengeChanged(stepIndex: stepIndex, stepLabel: stepLabel)
    }
  }

  public func onLivenessPassed(imageData: Data) {
    DispatchQueue.main.async { [weak self] in
      self?.delegate?.onLivenessPassed(imageData: imageData)
    }
  }

  public func onFailure(reason: String) {
    DispatchQueue.main.async { [weak self] in
      self?.delegate?.onFailure(reason: reason)
    }
  }

  public func onFaceDetected(boundingBox: CGRect?) {
    delegate?.onFaceDetected(boundingBox: boundingBox)
  }

  public func onFaceInOval(inside: Bool, reason: String?) {
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      self.overlayView.setFaceInOval(inside)
      if self.config.showInstructions {
        self.posHintLabel.isHidden = inside
        self.posHintLabel.text = reason ?? ""
      }
      self.delegate?.onFaceInOval(inside: inside, reason: reason)
    }
  }
}

File: android/liveness/src/main/java/com/liveness/detection/LivenessView.kt
package com.liveness.detection

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.camera.view.PreviewView
import androidx.lifecycle.LifecycleOwner

/**
 * Embeddable liveness view. Drop into any layout (XML or programmatic), assign a
 * [listener], optionally tune [config]/[sounds]/[modelUrl], then call [start].
 *
 * - Match `match_parent` to take over the screen.
 * - Wrap a sized container (e.g. a circular slot) to embed the camera in your own UI;
 *   in that case set `config.showInstructions = false` so the host owns text and buttons.
 *
 * Camera permission must be granted by the host before calling [start].
 */
class LivenessView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
  defStyleAttr: Int = 0,
) : FrameLayout(context, attrs, defStyleAttr), LivenessListener {

  private val density = context.resources.displayMetrics.density

  private val previewView: PreviewView = PreviewView(context).apply {
    layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
  }
  private val overlayView: LivenessOvalOverlayView = LivenessOvalOverlayView(context).apply {
    layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
  }
  private val hintView: LivenessHintView = LivenessHintView(context).apply {
    layoutParams = LayoutParams((52 * density).toInt(), (52 * density).toInt(), Gravity.CENTER_HORIZONTAL).apply {
      topMargin = (180 * density).toInt()
    }
  }
  private val instructionText: TextView = TextView(context).apply {
    layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT, Gravity.CENTER_HORIZONTAL).apply {
      topMargin = (220 * density).toInt()
    }
    text = "Position your face in the oval"
    setTextColor(Color.WHITE)
    textSize = 17f
    typeface = Typeface.DEFAULT_BOLD
    setShadowLayer(8f, 0f, 1f, 0x80000000.toInt())
  }
  private val posHintText: TextView = TextView(context).apply {
    layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT, Gravity.CENTER_HORIZONTAL).apply {
      topMargin = (260 * density).toInt()
    }
    setTextColor(Color.parseColor("#ff3b3b"))
    textSize = 13f
    gravity = Gravity.CENTER
    visibility = View.GONE
  }

  private var detector: LivenessDetector? = null

  /** Tuning + UI options. Re-assign at any time to update theme/shape/visibility. */
  var config: LivenessConfig = LivenessConfig()
    set(value) {
      field = value
      applyConfig()
    }

  /** Per-step sounds (left/blink/right/nod/mouth/good/capture). */
  var sounds: LivenessSoundOptions? = null

  /** Override the MediaPipe Face Landmarker model URL. Defaults to Google's hosted float16 model. */
  var modelUrl: String? = null

  /** Callbacks for the host. Will be invoked on the main thread. */
  var listener: LivenessListener? = null

  init {
    addView(previewView)
    addView(overlayView)
    addView(hintView)
    addView(instructionText)
    addView(posHintText)
    applyConfig()
  }

  private fun applyConfig() {
    overlayView.applyStyle(config)
    overlayView.setStepCount(LivenessStep.resolve(config.steps).size)
    val show = config.showInstructions
    instructionText.visibility = if (show) View.VISIBLE else View.GONE
    hintView.visibility = if (show) View.VISIBLE else View.GONE
    if (!show) posHintText.visibility = View.GONE
    posHintText.setTextColor(config.progressErrorColor)
  }

  /** Number of steps the current session will run (after applying `config.steps`). */
  val sessionStepCount: Int
    get() = LivenessStep.resolve(config.steps).size

  /**
   * Begin liveness. Camera permission must be granted by the host.
   * [lifecycleOwner] binds CameraX — typically the host Activity or Fragment.
   */
  fun start(lifecycleOwner: LifecycleOwner) {
    stop()
    val url = modelUrl ?: ModelDownloader.DEFAULT_MODEL_URL
    if (config.showInstructions) instructionText.text = "Downloading model..."

    ModelDownloader.downloadIfNeeded(
      context = context,
      url = url,
      fileName = "face_landmarker.task",
      maxAttempts = config.cdnMaxRetries,
      attemptTimeoutMs = config.cdnAttemptTimeoutMs.toInt(),
      connectivityCheckTimeoutMs = config.connectivityCheckTimeoutMs.toInt(),
      onSuccess = { file ->
        post {
          if (config.showInstructions) instructionText.text = "Position your face in the oval"
          val det = LivenessDetector(context, this, config, sounds)
          detector = det
          det.startLiveness(lifecycleOwner, previewView, true, ModelSource.FilePath(file.absolutePath))
        }
      },
      onError = { error ->
        post { listener?.onFailure(error) }
      }
    )
  }

  /** Stop detection and release the camera. Safe to call repeatedly. */
  fun stop() {
    detector?.stop()
    detector = null
  }

  // ── LivenessListener (internal — updates UI then forwards to [listener]) ──

  override fun onChallengeChanged(stepIndex: Int, stepLabel: String) {
    post {
      if (config.showInstructions) instructionText.text = stepLabel
      val total = sessionStepCount
      if (stepIndex >= 0) {
        overlayView.setProgress(stepIndex)
        overlayView.setStepDots(stepIndex)
        hintView.setHint(stepLabel)
      } else {
        overlayView.setProgress(total)
        overlayView.setStepDots(total)
        hintView.setHint(null)
      }
      listener?.onChallengeChanged(stepIndex, stepLabel)
    }
  }

  override fun onFaceInOval(inside: Boolean, reason: String?) {
    post {
      overlayView.setFaceInOval(inside)
      if (config.showInstructions) {
        posHintText.visibility = if (inside) View.GONE else View.VISIBLE
        posHintText.text = reason ?: ""
      }
      listener?.onFaceInOval(inside, reason)
    }
  }

  override fun onLivenessPassed(imageBytes: ByteArray) {
    post { listener?.onLivenessPassed(imageBytes) }
  }

  override fun onFailure(reason: String) {
    post { listener?.onFailure(reason) }
  }

  override fun onFaceDetected(boundingBox: android.graphics.RectF?) {
    listener?.onFaceDetected(boundingBox)
  }
}
